### DISCUSSION TITLE
Mirror codebase structure in tests

### DISCUSSION THREAD

**Discussion by ricardoV94** [pymc-devs/pymc3#5579] [General]
I think the tests files should mirror the original codebase. So they should be in a file with the same name as where the functions/classes that it tests are defined. E.g.,, timeseries tests should be in tests/distributions/test_timeseries.py. This is how Aesara does it and its much easier to find what tests relate to what code, and also to spot gaps in our tests.

_Originally posted by @ricardoV94 in https://github.com/pymc-devs/pymc/pull/5298#discussion_r822280003_

This would also allow to break up the immensely large and slow `test_distributions.py`, which was already brought up in https://github.com/pymc-devs/pymc/issues/5410#issuecomment-1029759988

Also, https://github.com/pymc-devs/pymc/pull/5577#discussion_r824542298

**Comment by twiecki** [comment]
Yeah, that makes a lot more sense.

**Comment by michaelosthege** [comment]
Related, here's a PR to combine the workflows into one: https://github.com/pymc-devs/pymc/pull/5623

Because our tests take so much time, it doesn't make sense to run a dense matrix in the workflows. Our current one is already sparse:

![image](https://user-images.githubusercontent.com/5894642/159126846-b74c07aa-ed55-4c06-82c6-9bb069e2c4f5.png)

But the definition in YAML is a little clumsy, with some things parametrized via the `strategy/matrix` and others via manually managed groups of test files.
What if we changed that into something that's A) easier to manage and B) more flexible?
It would be nice if we could just "tick boxes" in the above table.
It could be done by slicing that information out of the workflow YAML. For example having a spreadsheet, CSV, JSON or similar file that figures out which test files should run in that context.
The script in the workflow job could then get the pytest arguments from a bash/bat/py script.

__EDIT:__ After learning that line breaks in the `test-subset` are problematic on Windows, it's even more attractive to define this info outside of YAML.
